packages/app/ios/Orbit/IngestEventEmitter.m

```diff
@@ -0,0 +1,78 @@
+//
+//  EventEmitter.m
+//  Orbit
+//
+//  Created by Ozzie Kirkby on 2022-05-07.
+//
+
+#import <Foundation/Foundation.h>
+
+#import "IngestEventEmitter.h"
+#import <libkern/OSAtomic.h>
+#import <stdatomic.h>
+
+@implementation IngestEventEmitter
+
+@synthesize semaphore;
+
+- (instancetype)init {
+  self = [super init];
+  return self;
+}
+
++ (BOOL)requiresMainQueueSetup {
+    return NO;
+}
+
+-(void)startObserving {
+    self.hasListeners = YES;
+}
+
+// Will be called when this module's last listener is removed, or on dealloc.
+-(void)stopObserving {
+    self.hasListeners = NO;
+}
+
+RCT_EXPORT_MODULE();
+
++ (id)allocWithZone:(NSZone *)zone {
+    static IngestEventEmitter *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [super allocWithZone:zone];
+    });
+    return sharedInstance;
+}
+
+- (NSArray<NSString *> *)supportedEvents {
+    return @[@"onIngestEvent"];
+}
+
+- (bool) emitIngestEvent:(NSString *) fileJSON {
+  // update state
+  self.intentState = kWaiting;
+  
+  // emit event to RN
+  [self sendEventWithName:@"onIngestEvent" body:@{@"json": fileJSON}];
+  
+  // block until response from RN
+  dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+  self.semaphore = semaphore;
+  dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+  self.semaphore = nil;
+  
+  // return status
+  return [self intentState] == kSuccess;
+}
+
+RCT_EXPORT_METHOD(completedIngestion:(BOOL *)success) {
+  if (success) {
+    self.intentState = kSuccess;
+  } else {
+    self.intentState = kFailure;
+  }
+  dispatch_semaphore_signal([self semaphore]);
+}
```

Comment:
Let me know if you know of a better way to structure this, but the problem is this: `emitIngestEvent` needs to return if an ingest event was successful (i.e was it able to ingest correctly), however React Native's `sendEventWithName` happens async. To get around this, I setup this blocking semaphore which is resolved when the static  `completedIngestion` is called. This of course a bit fragile to `completedIngestion` always being called, but it seemed like the most reasonable work around I could find.

Reply:
Oof, yikes. Yeah, I've run into this kind of problem before, and I've had to do this kind of thing.

Happily, in this case, it looks like you may be able to dodge the problem. AFAICT `emitIngestEvent` is only used in `handleIngest:completion:`, but that method's already set up to be async, right? Can you call the completion handler from there after a response comes back?